### ISSUE TITLE
Enhance the link macro to support linking to a relative AsciiDoc document

### ISSUE THREAD

**Issue by mojavelinux** [asciidoctor/asciidoctor#417]
Enhance the link macro so it can be used to link to a relative AsciiDoc document. This is a special type of link because the target extension needs to be adjusted according to the backend.
- When the source document is rendered as HTML, the link should point to another HTML document.
- When the source document is rendered to DocBook, the link should point to another DocBook document. 
- Ideally, when the source document is displayed on GitHub, the link should reference the other AsciiDoc document (i.e., the target is unaltered).

Given

```
link:other-document.asciidoc[]
```

When rendered as HTML, the expected output is:

```
<a href="other-document.html">other-document.html</a>
```

When rendered as DocBook, the expected output is:

```
<ulink url="other-document.xml">other-document.xml</ulink>
```

When rendered on GitHub, the expected output is:

```
<a href="other-document.asciidoc">other-document.asciidoc</a>
```

The link macro may require the use of an extra hint (e.g., attribute) that activates this feature.

**Comment by nyeates** [comment]
I too had to pull in the
```
ifdef::env-github[:outfilesuffix: .adoc]
```

Wish it was in docs at http://asciidoctor.org/docs/asciidoc-writers-guide/

**Comment by mojavelinux** [comment]
Ironically, I had to do this too recently. See https://github.com/asciidoctor/asciidoctor-pdf/blame/master/README.adoc#L26-L28. I think we're still trying to figure out what the right solution is because GitHub throws us a curveball.

The correct place for the docs would be http://asciidoctor.org/docs/user-manual

I've filed an issue. See https://github.com/asciidoctor/asciidoctor.org/issues/398.

**Comment by pielgrzym** [comment]
The xref doesn't seem to work if we try to reference a document in subfolder:
xref:subfolder/document#[this should change the suffix to output format]

**Comment by mojavelinux** [comment]
This limitation has been addressed in the upcoming 1.5.3 release. See #844.

**Comment by pielgrzym** [comment]
I see, thanks for rapid response!

**Comment by zaxebo1** [comment]
@mojavelinux 
just for curiosity, any roughly scheduled date for 1.5.3 release?

**Comment by mojavelinux** [comment]
@zaxebo1 Imminent. As soon as I can get the issues done. I'm hoping by the beginning of August, but that's all depends on how quickly I can solve the outstanding issues. I want it out as much as anyone else does, trust me on that.

**Comment by oncletom** [comment]
Is it supported in a released version of Asciidoctor? It would be amazing and very useful :-)

**Comment by mojavelinux** [comment]
@oncletom Yes, this is available in Asciidoctor today. Note that there are some fixes to this feature that are waiting on the 1.5.3 release (notably #844).

**Comment by oncletom** [comment]
@mojavelinux ah cool! I indeed meant relative paths — same folder references work okay :-)

**Comment by axd1967** [comment]
as a result, this page should be deleted as it can only sow confusion:

https://github.com/asciidoctor/asciidoctor/wiki/How-to-link-to-a-relative-AsciiDoc-document

**Comment by mojavelinux** [comment]
I have updated the wiki page to point to the new section in the User Manual. I decided to leave the previous solution for reference, but I make it clear it's no longer recommended.

**Comment by chenrui333** [comment]
I am reading this thread, but it seems still unclear to me that does asciidoc in github supports something like `[link](subfolder/file)` this in markdown?

**Comment by madegroot** [comment]
For everyone searching for the solution for this: 
`<<page2.adoc#,Smart Link to page 2>>`
e.g. for backend HTML, this will be rendered to:
`<a href="page2.html">Smart Link to page 2</a>`

**Comment by mojavelinux** [comment]
FYI, as of 1.5.7 the trailing `#` is not required when the path ends with `.adoc`.

**Comment by whnr** [comment]
@mojavelinux 
> FYI, as of 1.5.7 the trailing `#` is not required when the path ends with `.adoc`.

I'm running 1.5.8 and omitting the `#` does not work.
`<<TODO.adoc,TODO page>>` becomes ` <a href="#TODO.adoc">TODO page</a>`

**Comment by mojavelinux** [comment]
...only when you're using the formal macro.

```
xref:TODO.adoc[TODO page]
```

**Comment by manoelcampos** [comment]
Sorry if this was already discussed. The `:outfilesuffix:` attribute works fine for browser and GitHub, but having to add the variable `{outfilesuffix}` to the end of each link for an AsciiDoc file is not intuitive.

That is what I think would be an ideal solution (I have no idea of the complexity or implications):

1. I keep hardcoding the`.adoc` extension in my links: `link:my-file.adoc[my file]`.
2. I shouldn't be required to add the `{outfilesuffix}` variable. If there is a hardcoded extension, the compiler would automatically change it to the defined `:outfilesuffix:` attribute.
3. If the `{outfilesuffix}` is present into a link, the compiler just replaces with its value.

I'd love to hear what you think.

**Comment by mojavelinux** [comment]
Please don't introduce new ideas / questions into an old issue.

> but having to add the variable `{outfilesuffix}` to the end of each link for an AsciiDoc file is not intuitive.

No one is saying this is required. To make a link to another AsciiDoc document, you use the xref macro, not the link macro.

```
xref:TODO.adoc[TODO page]
```

The outfilesuffix will be computed automatically based on the environment.

**Comment by manoelcampos** [comment]
> Please don't introduce new ideas / questions into an old issue.

Sorry, next time I find an open discussion or create new issue.
Thanks for the reply.

**Comment by mojavelinux** [comment]
Thanks!
